ext/java/org/msgpack/jruby/Packer.java

```diff
@@ -30,20 +35,139 @@ public IRubyObject allocate(Ruby runtime, RubyClass type) {
     }
   }
 
+  static class ExtRegistry {
+    private Ruby runtime;
+    public RubyHash hash;
+    public RubyHash cache;
+
+    public ExtRegistry(Ruby runtime) {
+      this.runtime = runtime;
+      hash = RubyHash.newHash(runtime);
+      cache = RubyHash.newHash(runtime);
+    }
+
+    public ExtRegistry dup() {
+      ExtRegistry copy = new ExtRegistry(this.runtime);
+      copy.hash = (RubyHash) this.hash.dup(runtime.getCurrentContext());
+      copy.cache = RubyHash.newHash(runtime);
+      return copy;
+    }
+
+    public void put(RubyClass klass, int typeId, IRubyObject proc, IRubyObject arg) {
+      RubyArray e = RubyArray.newArray(runtime, new IRubyObject[] { RubyFixnum.newFixnum(runtime, typeId), proc, arg });
```

Comment:
These are used by `registered_types_internal` method for `#registered_types` and `#type_registered?` methods of `Factory`, `Packer` and `Unpacker` defined under `lib/msgpack`.
Using RubyHash/RubyArray is just for implementation efficiency. Of course we can convert Java objects into RubyHash/RubyArrah in `registered_types_internal` if you want to do it when refactoring.